packages/hawtio/src/plugins/jmx/JmxContent.tsx

```diff
@@ -84,7 +87,17 @@ export const JmxContent: React.FunctionComponent = () => {
             </Routes>
           </React.Fragment>
         )}
-        {navItems.length === 0 && !selectedNode.objectName && <JmxContentMBeans />}
+        {navItems.length === 0 && !selectedNode.objectName && mBeanCollectionApplicable(selectedNode) && (
+          <AttributeTable />
+        )}
+        {navItems.length === 0 &&
+          !selectedNode.objectName &&
+          !mBeanCollectionApplicable(selectedNode) &&
+          !selectedNode.parent && <NodeNameTable />}
+        {navItems.length === 0 &&
+          !selectedNode.objectName &&
+          !mBeanCollectionApplicable(selectedNode) &&
+          selectedNode.parent && <JmxContentMBeans />}
```

Comment:
Also, I don't still understand why we need these three components `JmxContentMBeans`, `AttributeTable`, and `NodeNameTable`. Can't `JmxContentMBeans` be just implemented as `AttributeTable`?  What was the legacy Hawtio's behaviour when it had mbeans and folders at the same time?

Reply:
In my opinion, doing so would just make the code far, far more unreadable. Look at the amalgamation that handles the rendering. It takes over 140 lines, mixes different use cases, and doesn't even have all the logic at all (as you need the defined COLUMN_DEFS to get some of the logic out, and a lot of those if-elses are just plain different use cases mixed together. In comparison, this only spans 12 and every use case is clearly defined in its own component.

This way, the filtering order is really clear in how a component is chosen, and every component has its own separate logic to handle the nodes.

```    function render(response: { request: any, value: any }): void {
      let data = response.value;
      let mbeanIndex = $scope.mbeanIndex;
      let mbean = response.request['mbean'];

      if (mbean) {
        // lets store the mbean in the row for later
        data["_id"] = mbean;
      }
      if (mbeanIndex) {
        if (mbean) {

          let idx = mbeanIndex[mbean];
          if (!angular.isDefined(idx)) {
            idx = $scope.mbeanRowCounter;
            mbeanIndex[mbean] = idx;
            $scope.mbeanRowCounter += 1;
          }
          if (idx === 0) {
            // this is to force the table to repaint
            $scope.selectedIndices = $scope.selectedItems.map((item) => $scope.gridData.indexOf(item));
            gridData = [];

            if (!$scope.columnDefs.length) {
              // lets update the column definitions based on any configured defaults

              let key = workspace.selectionConfigKey();
              let defaultDefs = _.clone(workspace.attributeColumnDefs[key]) || [];
              let defaultSize = defaultDefs.length;
              let map = {};
              angular.forEach(defaultDefs, (value, key) => {
                let field = value.field;
                if (field) {
                  map[field] = value
                }
              });

              let extraDefs = [];
              _.forEach(data, (value, key) => {
                if (includePropertyValue(key, value)) {
                  if (!map[key]) {
                    extraDefs.push({
                      field: key,
                      displayName: key === '_id' ? 'Object name' : Core.humanizeValue(key),
                      visible: defaultSize === 0
                    });
                  }
                }
              });

              // the additional columns (which are not pre-configured), should be sorted
              // so the column menu has a nice sorted list instead of random ordering
              extraDefs = extraDefs.sort((def, def2) => {
                // make sure _id is last
                if (_.startsWith(def.field, '_')) {
                  return 1;
                } else if (_.startsWith(def2.field, '_')) {
                  return -1;
                }
                return def.field.localeCompare(def2.field);
              });
              extraDefs.forEach(e => defaultDefs.push(e));

              if (extraDefs.length > 0) {
                $scope.hasExtraColumns = true;
              }

              $scope.columnDefs = defaultDefs;
            }
          }
          // mask attribute read error
          _.forEach(data, (value, key) => {
            if (includePropertyValue(key, value)) {
              data[key] = maskReadError(value);
            }
          });
          // assume 1 row of data per mbean
          gridData[idx] = data;

          let count = $scope.mbeanCount;
          if (!count || idx + 1 >= count) {
            // only cause a refresh on the last row
            let newSelections = $scope.selectedIndices.map((idx) => $scope.gridData[idx]).filter((row) => row);
            $scope.selectedItems.splice(0, $scope.selectedItems.length);
            $scope.selectedItems.push.apply($scope.selectedItems, newSelections);
            $scope.gridData = gridData;
            Core.$apply($scope);
          }
          // if the last row, then fire an event
        } else {
          log.info("No mbean name in request", JSON.stringify(response.request));
        }
      } else {
        $scope.columnDefs = PROPERTIES_COLUMN_DEFS;
        let showAllAttributes = true;
        if (_.isObject(data)) {
          let properties = [];
          _.forEach(data, (value, key) => {
            if (showAllAttributes || includePropertyValue(key, value)) {
              // always skip keys which start with _
              if (!_.startsWith(key, "_")) {
                // lets format the ObjectName nicely dealing with objects with
                // nested object names or arrays of object names
                if (key === "ObjectName") {
                  value = unwrapObjectName(value);
                }
                // lets unwrap any arrays of object names
                if (_.isArray(value)) {
                  value = value.map((v) => unwrapObjectName(v));
                }
                // the value must be string as the sorting/filtering of the table relies on that
                let type = lookupAttributeType(key);
                let data = {
                  key: key,
                  name: Core.humanizeValue(key),
                  value: maskReadError(Core.safeNullAsString(value, type))
                };

                generateSummaryAndDetail(key, data);
                properties.push(data);
              }
            }
          });
          if (!_.some(properties, (p) => {
            return p['key'] === 'ObjectName';
          })) {
            let objectName = {
              key: "ObjectName",
              name: "Object Name",
              value: mbean
            };
            generateSummaryAndDetail(objectName.key, objectName);
            properties.push(objectName);
          }
          properties = _.sortBy(properties, 'name');
          $scope.selectedItems = [data];
          data = properties;
        }
        $scope.gridData = data;
        Core.$apply($scope);
      }
    }
    ```

Reply:
I don't think you answered my question. I wasn't asking about how the legacy hawtio is implemented, but asking about what it behaves in the same cases. What does each of JmxContentMBeans, AttributeTable, and NodeNameTable do?  Why do we need JmxContentMBeans and AttributeTable at the same time?  What was the legacy Hawtio's behaviour when it had mbeans and folders at the same time?

Reply:
Each node had configured a view in a per node basis from what I gathered (for example some folders like the topmost jmx one in the tree view had the tabular, while the others had the node names or just the default with objectName)

I tried to find a common logic that decided how each one was configured (specially because we are missing some nodes like the ones in camel) and from what I have been testing they match for almost all cases

Reply:
On top of that each table is a different view

Attributes show a list of attribute-valur pair with the mbean node attributes

AttributesTable show a list of the attributes for every mbean in a row that are children of the node.

NodeName is used when not all nodes below are mbeans. Seeing as this behaviour only appeared in parent nodes i went with that logic as any one that was not a parent node usually had the JMXContentMBean default one. It shows the name of the children nodes

JMXContentMBean is the default view, which only shows the objectNames

I tried to be as close to what the other version was showing

Reply:
OK, so what I still don't see yet is the difference between `NodeNameTable` and `JMXContentMBeans`. Why do we need the two instead of one?  You said:

> JMXContentMBean is the default view, which only shows the objectNames

it suggests that the node is actually a parent, as otherwise it shouldn't have a set of objectNames. So `NodeNameTable` is for a parent but the default view is also for a parent, right?  I guess then we really don't yet clearly identify the true conditions to dispatch each view.

Apart from code, let's think it logically. I think there are four cases to capture in JMX view:

1. Single MBean
2. Multiple MBeans of the same kind
3. Multiple MBeans of different kinds
4. Mixture of MBeans and folder nodes

Are they MECE?  Then `NodeNameTable` corresponds to 4 and the default to 3, right?  If so, probably the conditions you defined haven't captured those switches clearly, because the difference between 3 and 4 is not whether it's parent or not.

Reply:
Case 3 with NodeName

![image](https://user-images.githubusercontent.com/17336236/230022720-74aa09fb-e4d7-4cff-9bb0-5de52403a600.png)

Case 4 with ObjectName which is the default view

![image](https://user-images.githubusercontent.com/17336236/230022647-a5a0549d-0db7-4552-9409-8da0cf5e5966.png)

Case 4 happens when is a folder that is not  a topmost parent if that makes thing more clear. Actually it happens if no other conditions are met according to the other version, as I pasted before.

I can also work on a simplified version of the old code if that's easier...? I mean, at this point I'm not sure what's the problem

Reply:
Sorry if it confuses you, but in this case the behaviours of the legacy version might be a bit buggy, at least it's not consistent. I think we should think up a clearer specification now.

In case 4, why do we need different views?  I don't think it makes sense to distinguish a top-most parent from the other parents. For case 4, I think we should show object names for mbeans but just show the names for folders. So it's what `JMXContentMBeans` does, I think.

> Case 2 and 3 are kind of the same and use the same view (tabular attribute table)

Hmm, then how can the table shows MBeans that have different sets of attributes?  I thought it has a clear difference between the two.

So, to wrap up, now I think we only need three views:

1. `Attributes` for case 1
2. `AttributesTable` for case 2
3. `JMXContentMBeans` for case 3 & 4

WDYT?

Reply:
>    Case 2 and 3 are kind of the same and use the same view (tabular attribute table)

> Hmm, then how can the table shows MBeans that have different sets of attributes? I thought it has a clear difference 
> between the two.

It defaults to the nodeNameTable if it doesn't match (or now the JMXContentMBeans table), or if there is no valid information.

Alright about the other thing. I only have to remove the new view and the conditional and it should be fine.

Reply:
Ok so there is no current examples in current HawtIO to test case 3 (different kind of MBeans). I have checked all current nodes that are shown and every case always bring the same key structure so far. I'll put the code in place anyway to default to JMXContentMBeans when they don't match.

Also for the case where some MBeans don't bring info at all it also defaults to the JMXContentMBeans view
![image](https://user-images.githubusercontent.com/17336236/230053597-08953bf6-9439-409d-a735-c98f727461bf.png)
![image](https://user-images.githubusercontent.com/17336236/230053641-e125b9f8-636c-4012-a3ac-6a5318e28621.png)